### PR TITLE
[s11n 7/7] Implement context serialisation

### PR DESCRIPTION
**Pull request by jarlamsa**
## Description

This is a side-port to development of PR https://github.com/ARMmbed/mbedtls-restricted/pull/616 that has already been merged to baremetal.

**Dependencies:** based on #2764, #2766 #2785 #2788 #2789 and #2791  - see "Background" below.

This PR implements the full SSL context serialisation.

## Background

This is the seventh and final in a series of seven PRs implementing serialization of a full SSL context (not just the session), under some restrictions.

The PRs in the series are:
 
1. [r563](https://github.com/ARMmbed/mbedtls-restricted/pull/563) -> #2764 Refactor derive keys
2. [r568](https://github.com/ARMmbed/mbedtls-restricted/pull/568) -> #2766 Make session serialisation functions public 
3. [r570](https://github.com/ARMmbed/mbedtls-restricted/pull/570) -> #2785 Make session format ABI-independent
4. [r567](https://github.com/ARMmbed/mbedtls-restricted/pull/567) -> #2788 Add version + config check to SSL tickets
5. [r572](https://github.com/ARMmbed/mbedtls-restricted/pull/572) -> #2789 Context serialisation: API and documentation
6. [r575](https://github.com/ARMmbed/mbedtls-restricted/pull/575) -> #2791 Serialisation test 
7. [r616](https://github.com/ARMmbed/mbedtls-restricted/pull/616) -> #2792 Implement context serialisation __(this PR)__

The full series is already merged in baremetal. It needs upstreaming for the benefit of the Could team, who will be the main user, and relies on upstream releases.

## Side-porting details

Side-porting was done with `git cherry-pick`; some commits needed to be manually reworked after cherry-picking and some commits needed to be dropped because they didn't make sense in upstream.

The table shows how each side-ported commit was handled:

- (nothing) cherry-picking worked without conflict
- "t" only trivial context-related conflicts (eg. independent additions to the same place)
- "r" manual rework after cherry-picking
- "R" the commit was reconstructed manually (moving code around)
- "+" added in this side-port PR

| resolution | original commit | commit message |
|---| ---          | --- |
| | a3024eef7be4 | Save Hello random bytes for later use |
| r | 569ed6ba56e8 | Implement usage checks in context_save() |
| | b6163ef17558 | Document internal serialisation format |
| r | 5e534baaecbf | Add usage checks in context_load() |
| | d0dd10469bb0 | Add (stub) header writing and checking |
| | f1f3e529a5a0 | Add session saving/loading |
| | 3b23c7d2d21c | Improve demo/testing code in client/server2 |
| | 8175816200ef | Fix English in comments |
| r | 322f3c73771e | Add transform (de)serialization |
| r | c981229b04cd | Fix memory leak in client/server2 |
| | 16d1485a3d49 | Add saved fields from top-level structure |
| | 138079d7d6d7 | Add setting of forced fields when deserializing |
| r | 86dfa0cfc954 | Change requirements for setting timer callback |
| | 0d83271a45ef | Enable serialisation tests in ssl-opt.sh |
| | ff22200fab6a | Re-use buffer allocated by handshake_init() |
| | bc847caa333c | Actually reset the context on save as advertised |
| | 2f3fa62a0afb | Fix compiler warning: comparing signed to unsigned |
| - | 73a4636ca4eb | Adapt to hardcoded single version |
| | 4c1d06e429f4 | Provide serialisation API only if it's enabled |
| | a7cd4830ee56 | Implement config-checking header to context s11n |
| | 7ce9446e4c32 | Avoid duplication of session format header |
| r | 2cc9223a3bdb | Fix compile error in reduced configurations |
| | 7af73754739c | Fix MSVC warning |
| | 14e2a8ac06b3 | Fix a typo in a comment |
| - | b3bb31bd90b1 | Introduce getter function for disable_renego |
| - | 18332c5c6cdd | Improve getter for renegotiation enabled |
| r | 69a3e417d8e7 | Improve reability and debugability of large if |
| + | e1da74e980db | Add missing guards for mac usage |

## Review notes

Reviewers are strongly encouraged to review commit by commit, as the final diff is unlikely to make sense. Please note that the first commits (up to "Fix style issues") are part of #2764, #2766 #2785 #2788 #2789 and #2791 and don't need to be re-reviewed here (except checking that they're indeed part of #2764 #2766 #2785 #2788 #2789 #2791).

### CODE REVIEW

**Comment by hanno-arm** [comment]
@jarlamsa There are CI failures which appear to result from the use of ciphersuite getter functions which haven't been sideported yet. Could you have a look?

**Comment by jarlamsa** [comment]
> @jarlamsa There are CI failures which appear to result from the use of ciphersuite getter functions which haven't been sideported yet. Could you have a look?

Worked through those and amended the commits that brought up issues and fixed those.

**Comment by jarlamsa** [comment]
Fixed an issue with missing guards when no ciphersuite is using mac

File: programs/ssl/ssl_server2.c [hunk 1]
```diff
@@ -3914,7 +3914,6 @@ int main( int argc, char *argv[] )
                             opt.nbio == 0 ? mbedtls_net_recv_timeout : NULL );
 
 #if defined(MBEDTLS_TIMING_C)
-            if( opt.nbio != 0 && opt.read_timeout != 0 )
                 mbedtls_ssl_set_timer_cb( &ssl, &timer,
```
Comment (hanno-arm) [inline programs/ssl/ssl_server2.c]: Minor: The indentation is off here.

File: include/mbedtls/ssl.h [hunk 2]
```diff
@@ -3736,6 +3829,130 @@ int mbedtls_ssl_close_notify( mbedtls_ssl_context *ssl );
...
+ *                 state. You're then free to either re-use the context
+ *                 structure for a different connection, or call
+ *                 mbedtls_ssl_free() on it. See the documentation of
+ *                 mbedtls_ssl_session_reset() for more details.
+ *
+ * \param ssl      The SSL context to save. On success, it is no longer
+ *                 associated with the connection that has been serialized.
+ * \param buf      The buffer to write the serialized data to. It must be a
+ *                 writeable buffer of at least \p len bytes, or may be \c
+ *                 NULL if \p len is \c 0.
```
Comment (samhaa01) [inline include/mbedtls/ssl.h]: Is this referring to buf_len?

Comment (jarlamsa) [reply include/mbedtls/ssl.h]: Yes, good catch, this was introduced in https://github.com/ARMmbed/mbedtls/pull/2789

Comment (jarlamsa) [reply include/mbedtls/ssl.h]: Fixed in this PR.

**Comment by jarlamsa** [comment]
rebased

**Comment by jarlamsa** [comment]
@hanno-arm please re-review

**Comment by jarlamsa** [comment]
Added some debug-prints, @artokin please review the latest addition

**Comment by jarlamsa** [comment]
Agreed with @Patater that the ChangeLog entry for the record checking feature which was merged with #2790 is added here.

**Comment by jarlamsa** [comment]
@hanno-arm @artokin please re-review

**Comment by jarlamsa** [comment]
All of the serialisation PRs now have 2 approving reviews and passing CI. Thus considering all of them ready for merge.

**Comment by mpg** [comment]
Note: I've started a release job on this PR: https://jenkins-internal.mbed.com/blue/organizations/jenkins/mbedtls-release-new/detail/mbedtls-release-new/441/pipeline

**Pull request merged**